### PR TITLE
Handle edge case of form direct access policy

### PR DESCRIPTION
**Pull request by AdrienClairembault**
## Checklist before requesting a review

- [X] I have performed a self-review of my code.
- [ ] I have added tests that prove my fix is effective or that my feature works.
   -> Not easy to add a good cypress tests for this, it would take too much execution time for such a rare edge case that has no securities concerns

## Description

As a reminder, the direct access policies allow a form to be access with a token in the url.
This allow a user to access the form page, and at this point we will store the token in session to avoid having to add it to every AJAX request done on the form page (since the user used the correct token once, it is safe to assume he can keep access to the form for the rest of his session).

There was a small edge case as the form was then visible from the service catalog (since the token was still in the session).

I've switched to another mechanism where I check for the token in the referer url, this way we can cover the case of AJAX requests on the form page + explicitly blacklist the service catalog page from the fallback to handle the extra rare case of an user going straight to the service catalog after being on the form page.

Note that there are no security issues, this is just to avoid showing the form in the service catalog but the user had access to it anyway since he knows the token.

## References

Internal support ticket: !39558

### CODE REVIEW

**Comment by trasher** [comment]
>     * [ ]  I have added tests that prove my fix is effective or that my feature works.
>       -> Not easy to add a good cypress tests for this, it would take too much execution time for such a rare edge case that has no securities concerns

What about a "simple" phpunit test?

File: src/Glpi/Form/AccessControl/ControlType/DirectAccess.php [hunk 1]
```diff
@@ -168,22 +170,33 @@ private function validateToken(
...
+            // Disable this fallback for the service catalog, this prevent the
+            // edge cases of users going straight to the service catalog after
+            // being on the form itself.
+            && !str_contains($current_route_path, "ServiceCatalog")
+        ) {
+            $parsed_referer = parse_url($referer);
+            $query = $parsed_referer['query'] ?? "";
+            parse_str($query, $parsed_query);
+            $token = $parsed_query['token'] ?? null;
+        }
```
Comment (cedric-anne) [inline src/Glpi/Form/AccessControl/ControlType/DirectAccess.php]: Relying on the referer may be an issue. Indeed, this is not a mandatory request header and some privacy focused browser extensions may remove all referer headers from web requests.

**Comment by AdrienClairembault** [comment]
Indeed, I guess I'll keep the original session solution but with a strict exclusion for the service catalog.
